File: sdks/ios/Sources/RelayMessenger/Relay.swift
import Foundation

/// The public entry point to the Relay messenger. A single shared actor holds session state
/// (token, config, theme) and serialises access to the API client.
///
/// Typical lifecycle:
/// ```
/// try await Relay.shared.boot(appId: "wrk_…")                 // anonymous lead
/// try await Relay.shared.login(externalId: "u_42",            // upgrade to identified user;
///                              userHash: hashFromYourBackend)  // user_hash is HMAC'd server-side
/// ```
public actor Relay {
    public static let shared = Relay()
    private init() {}

    public static let defaultBaseURL = URL(string: "https://api.relay.dev")!

    private var client: RelayClient?
    private var appId: String?
    private var contactValue: Contact?
    private var configValue: MessengerConfig?
    private var themeValue = RelayTheme()
    private var hostTheme: RelayTheme?          // host override wins over boot config
    private var lastDeviceTokenHex: String?
    private var pushEnvironment = "production"
    private let bundleId = Bundle.main.bundleIdentifier

    // MARK: - Boot & identity

    /// Boots an anonymous ("lead") session. Resolves the workspace from the public `app_id`.
    @discardableResult
    public func boot(appId: String,
                     baseURL: URL = Relay.defaultBaseURL,
                     theme: RelayTheme? = nil) async throws -> Contact {
        self.appId = appId
        if let theme { hostTheme = theme }
        self.client = RelayClient(baseURL: baseURL)
        return try await performBoot(user: nil, userHash: nil)
    }

    /// Identifies the current session as a known user.
    /// - Parameter userHash: `hex(HMAC_SHA256(workspace_identity_secret, externalId))`, computed
    ///   by the integrator's **backend**. The workspace secret is NEVER shipped in the app — the
    ///   SDK only ever forwards this precomputed hash.
    @discardableResult
    public func login(externalId: String,
                      userHash: String,
                      email: String? = nil,
                      name: String? = nil) async throws -> Contact {
        guard appId != nil else { throw RelayError.notBooted }
        return try await performBoot(user: BootUser(externalId: externalId, email: email, name: name),
                                     userHash: userHash)
    }

    private func performBoot(user: BootUser?, userHash: String?) async throws -> Contact {
        guard let client, let appId else { throw RelayError.notBooted }
        let resp = try await client.boot(BootRequest(appId: appId, user: user, userHash: userHash))
        await client.setToken(resp.sessionToken)
        contactValue = resp.contact
        configValue = resp.config
        themeValue = hostTheme ?? RelayTheme.from(config: resp.config)
        return resp.contact
    }

    /// Clears the local session and retires this device's push token (best effort).
    /// Call `boot()` again afterwards for a fresh anonymous session.
    public func logout() async {
        if let hex = lastDeviceTokenHex, let client { try? await client.unregisterDevice(token: hex) }
        lastDeviceTokenHex = nil
        contactValue = nil
        await client?.setToken(nil)
    }

    // MARK: - State accessors

    public func contact() -> Contact? { contactValue }
    public func config() -> MessengerConfig? { configValue }
    public func theme() -> RelayTheme { themeValue }
    public func setTheme(_ theme: RelayTheme) { hostTheme = theme; themeValue = theme }

    // MARK: - Conversations

    public func conversations(cursor: String? = nil, limit: Int? = nil) async throws -> Page<Conversation> {
        try await requireClient().conversations(cursor: cursor, limit: limit)
    }

    public func parts(conversationId: String, after: String? = nil,
                      cursor: String? = nil, limit: Int? = nil) async throws -> Page<Part> {
        try await requireClient().parts(conversationId: conversationId, cursor: cursor, after: after, limit: limit)
    }

    public func startConversation(body: String, attachments: [Attachment]? = nil) async throws -> Conversation {
        try await requireClient().createConversation(body: body, attachments: attachments)
    }

    /// Sends a reply. The client attaches a fresh `Idempotency-Key` so retries are safe.
    public func reply(conversationId: String, body: String, attachments: [Attachment]? = nil) async throws -> Part {
        try await requireClient().reply(conversationId: conversationId, body: body, attachments: attachments)
    }

    public func rate(conversationId: String, rating: Int, remark: String? = nil) async throws -> Part {
        try await requireClient().rate(conversationId: conversationId, rating: rating, remark: remark)
    }

    // MARK: - Push

    /// Registers this device's APNs token. The server upserts on the token, so it is safe —
    /// and expected — to call this on every launch and again on every APNs token refresh.
    public func registerForPushNotifications(deviceToken: Data, environment: String = "production") async throws {
        let hex = PushManager.hexToken(from: deviceToken)
        lastDeviceTokenHex = hex
        pushEnvironment = environment
        _ = try await requireClient().registerDevice(
            DeviceRegister(token: hex, appId: bundleId, environment: environment)
        )
    }

    public func unregisterPush() async throws {
        guard let hex = lastDeviceTokenHex else { return }
        try await requireClient().unregisterDevice(token: hex)
        lastDeviceTokenHex = nil
    }

    /// Pulls the deep-link `conversation_id` out of a notification payload on tap. `nonisolated`
    /// so it can be called synchronously from an `AppDelegate`/`UNUserNotificationCenterDelegate`.
    public nonisolated func handleNotification(userInfo: [AnyHashable: Any]) -> String? {
        PushManager.conversationId(from: userInfo)
    }

    // MARK: - Attachments

    /// Uploads a file (presign → PUT to S3) and returns the `Attachment` reference to include
    /// in `reply`/`startConversation`.
    public func uploadAttachment(data: Data, filename: String, contentType: String) async throws -> Attachment {
        let client = try requireClient()
        let presign = try await client.presign(filename: filename, contentType: contentType)
        try await client.putFile(to: presign.uploadUrl, data: data, contentType: contentType)
        return Attachment(key: presign.key, filename: filename, contentType: contentType)
    }

    // MARK: -

    private func requireClient() throws -> RelayClient {
        guard let client else { throw RelayError.notBooted }
        return client
    }
}

File: sdks/android/relay-messenger/src/main/kotlin/dev/relay/messenger/Relay.kt
package dev.relay.messenger

import android.content.Context
import android.content.Intent
import java.util.UUID

/**
 * Public entry point for the Relay messenger SDK. Singleton; all network calls are `suspend`.
 *
 * Lifecycle: [boot] once at launch (anonymous "lead"), optionally [login] to identify the user
 * with a backend-computed `user_hash`. The SDK never holds the workspace identity secret.
 *
 * Session token + last identity are persisted so a cold-started [ReplyReceiver] (direct reply
 * from a notification) can re-attach without a full app launch.
 */
object Relay {

    const val DEFAULT_BASE_URL = "https://api.relay.dev"

    /** Deep-link + push payload keys, matching the API contract's `data` payload. */
    const val KEY_CONVERSATION_ID = "conversation_id"
    const val KEY_PUSH_TYPE = "type"
    const val PUSH_TYPE_REPLY = "conversation.reply"

    private const val PREFS = "relay_messenger"
    private const val K_APP_ID = "app_id"
    private const val K_BASE = "base_url"
    private const val K_TOKEN = "session_token"
    private const val K_EXT_ID = "external_id"
    private const val K_HASH = "user_hash"
    private const val K_EMAIL = "email"
    private const val K_NAME = "name"
    private const val K_PUSH = "push_token"

    private lateinit var appContext: Context
    private lateinit var client: RelayClient

    /** Last boot config (theming source of truth) and resolved contact. */
    @Volatile var config: MessengerConfig? = null; private set
    @Volatile var contact: Contact? = null; private set

    val isBooted: Boolean get() = ::client.isInitialized && client.sessionToken != null

    private fun prefs() = appContext.getSharedPreferences(PREFS, Context.MODE_PRIVATE)

    // --- boot / identity ---

    /** Boot as an anonymous lead. Call once at app start with your public `wrk_…` app id. */
    suspend fun boot(
        context: Context,
        appId: String,
        baseUrl: String = DEFAULT_BASE_URL,
    ): BootResponse {
        appContext = context.applicationContext
        prefs().edit().putString(K_APP_ID, appId).putString(K_BASE, baseUrl).apply()
        client = RelayClient(baseUrl)
        return doBoot(externalId = null, userHash = null, email = null, name = null)
    }

    /**
     * Identify the current user. [userHash] is `HMAC_SHA256(secret, externalId)` computed by the
     * integrator's backend — the secret is never in the app.
     */
    suspend fun login(
        externalId: String,
        userHash: String,
        email: String? = null,
        name: String? = null,
    ): BootResponse {
        requireBooted()
        prefs().edit()
            .putString(K_EXT_ID, externalId)
            .putString(K_HASH, userHash)
            .putString(K_EMAIL, email)
            .putString(K_NAME, name)
            .apply()
        return doBoot(externalId, userHash, email, name)
    }

    /** Drop the identified session; unregisters the push token. Keeps app id + base url. */
    suspend fun logout() {
        requireBooted()
        prefs().getString(K_PUSH, null)?.let { runCatching { client.unregisterDevice(it) } }
        prefs().edit()
            .remove(K_TOKEN).remove(K_EXT_ID).remove(K_HASH)
            .remove(K_EMAIL).remove(K_NAME).remove(K_PUSH)
            .apply()
        client.sessionToken = null
        contact = null
    }

    private suspend fun doBoot(
        externalId: String?, userHash: String?, email: String?, name: String?,
    ): BootResponse {
        val user = if (externalId != null || email != null || name != null)
            BootUser(externalId, email, name) else null
        val resp = client.boot(
            BootRequest(appId = prefs().getString(K_APP_ID, null)!!, user = user, userHash = userHash),
        )
        client.sessionToken = resp.sessionToken
        contact = resp.contact
        config = resp.config
        prefs().edit().putString(K_TOKEN, resp.sessionToken).apply()
        // Re-register a cached push token against the (possibly new) contact.
        prefs().getString(K_PUSH, null)?.let { runCatching { registerPushToken(it) } }
        return resp
    }

    // --- conversations ---

    suspend fun conversations(cursor: String? = null, limit: Int? = null): Page<Conversation> =
        client.listConversations(cursor, limit)

    suspend fun parts(
        conversationId: String, after: String? = null, cursor: String? = null, limit: Int? = null,
    ): Page<Part> = client.listParts(conversationId, cursor, after, limit)

    suspend fun startConversation(body: String, attachments: List<Attachment>? = null): Conversation =
        client.createConversation(body, attachments)

    suspend fun reply(
        conversationId: String,
        body: String,
        attachments: List<Attachment>? = null,
        idempotencyKey: String = UUID.randomUUID().toString(),
    ): Part = client.reply(conversationId, body, attachments, idempotencyKey)

    suspend fun rate(conversationId: String, rating: Int, remark: String? = null): Part =
        client.rate(conversationId, rating, remark)

    // --- attachments ---

    /** Presign → PUT bytes → return the [Attachment] reference to include in a reply. */
    suspend fun uploadAttachment(filename: String, contentType: String, bytes: ByteArray): Attachment {
        val p = client.presign(filename, contentType)
        client.putBytes(p.uploadUrl, bytes, contentType)
        return Attachment(key = p.key, filename = filename, contentType = contentType)
    }

    // --- push ---

    /** Register/refresh the FCM token. Server upserts on the token, so this is safe every launch. */
    suspend fun registerPushToken(token: String) {
        requireBooted()
        prefs().edit().putString(K_PUSH, token).apply()
        client.registerDevice(
            DeviceRegister(platform = "android", token = token, appId = appContext.packageName),
        )
    }

    suspend fun unregisterPush(token: String? = null) {
        requireBooted()
        val t = token ?: prefs().getString(K_PUSH, null) ?: return
        client.unregisterDevice(t)
        prefs().edit().remove(K_PUSH).apply()
    }

    // --- push handling / deep link ---

    /**
     * Handle a Relay push `data` payload. Returns true if it was a Relay message (and a
     * notification was posted). Call from your own FirebaseMessagingService if you don't use
     * [RelayMessagingService].
     */
    fun handleRemoteMessage(context: Context, data: Map<String, String>): Boolean {
        val conversationId = data[KEY_CONVERSATION_ID] ?: return false
        postReplyNotification(context, conversationId, body = data["body"])
        return true
    }

    /** Extract the conversation id from a deep-link intent (notification tap). */
    fun conversationIdFrom(intent: Intent?): String? =
        intent?.getStringExtra(KEY_CONVERSATION_ID)

    /**
     * Send a reply on behalf of a (possibly cold-started) process, e.g. from [ReplyReceiver].
     * Re-attaches persisted state and re-boots from stored identity if the token is gone.
     */
    suspend fun replyFromNotification(context: Context, conversationId: String, text: String): Part {
        attach(context)
        if (client.sessionToken == null) reBootFromStored()
        return reply(conversationId, text)
    }

    /** Rebuild the client from persisted prefs (base url + token). Safe from any process. */
    @Synchronized
    fun attach(context: Context) {
        if (!::appContext.isInitialized) appContext = context.applicationContext
        if (!::client.isInitialized) {
            client = RelayClient(prefs().getString(K_BASE, DEFAULT_BASE_URL)!!)
        }
        client.sessionToken = prefs().getString(K_TOKEN, null)
    }

    private suspend fun reBootFromStored() {
        val extId = prefs().getString(K_EXT_ID, null)
        val hash = prefs().getString(K_HASH, null)
        doBoot(extId, hash, prefs().getString(K_EMAIL, null), prefs().getString(K_NAME, null))
    }

    private fun requireBooted() = check(::client.isInitialized) { "Relay.boot(...) must be called first" }
}
